### PR TITLE
Expressions with some failing tests

### PR DESCRIPTION
**Pull request by hakeemsm**
Removed grouping, highlighter and join per previous conversation.

### CODE REVIEW

File: msbuild.log [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+Build FAILED.
+
+       "C:\Projects\GitHub\lucene.net\Lucene.Net.sln" (default target) (1) ->
+       (ValidateSolutionConfiguration target) -> 
+         C:\Projects\GitHub\lucene.net\Lucene.Net.sln.metaproj : error MSB4126: The specified solution configuration "Release|x64" is invalid. Please specify a valid solution configuration using the Configuration and Platform properties (e.g. MSBuild.exe Solution.sln /p:Configuration=Debug /p:Platform="Any CPU") or leave those properties blank to use the default solution configuration. [C:\Projects\GitHub\lucene.net\Lucene.Net.sln]
+
+    0 Warning(s)
+    1 Error(s)
+
+Time Elapsed 00:00:00.91
```
Comment (synhershko) [inline msbuild.log]: please remove this file

File: src/Lucene.Net.Core/Lucene.Net.csproj [hunk 2]
```diff
@@ -40,9 +41,6 @@
     <StartupObject />
   </PropertyGroup>
   <ItemGroup>
-    <Reference Include="ICSharpCode.SharpZipLib">
-      <HintPath>..\..\packages\SharpZipLib.0.86.0\lib\20\ICSharpCode.SharpZipLib.dll</HintPath>
-    </Reference>
```
Comment (synhershko) [inline src/Lucene.Net.Core/Lucene.Net.csproj]: why did you remove this?

File: src/Lucene.Net.Core/packages.config [hunk 3]
```diff
@@ -1,5 +1,6 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
 <packages>
+  <package id="Antlr4.Runtime" version="4.3.0" targetFramework="net451" />
```
Comment (synhershko) [inline src/Lucene.Net.Core/packages.config]: why is this necessary in the core?

File: src/Lucene.Net.Tests.Facet/Lucene.Net.Tests.Facet.csproj [hunk 4]
```diff
@@ -34,7 +34,7 @@
       <HintPath>..\..\packages\Apache.NMS.1.6.0.3083\lib\net40\Apache.NMS.dll</HintPath>
     </Reference>
     <Reference Include="nunit.framework">
-      <HintPath>..\..\packages\NUnit.2.6.4\lib\nunit.framework.dll</HintPath>
+      <HintPath>..\..\packages\NUnit.2.6.3\lib\nunit.framework.dll</HintPath>
```
Comment (synhershko) [inline src/Lucene.Net.Tests.Facet/Lucene.Net.Tests.Facet.csproj]: why this change?

File: src/Lucene.Net.Tests.Classification/packages.config [hunk 5]
```diff
@@ -1,4 +0,0 @@
-﻿<?xml version="1.0" encoding="utf-8"?>
-<packages>
-  <package id="NUnit" version="2.6.4" targetFramework="net451" />
-</packages>
```
Comment (synhershko) [inline src/Lucene.Net.Tests.Classification/packages.config]: why was this file removed?

**Comment by synhershko** [comment]
Thanks, left some comments

Lets try to avoid touching unrelated files...

**Comment by hakeemsm** [comment]
OK, I'll fix them later this evening. Thx

**Comment by eladmarg** [comment]
great job!

File: src/Lucene.Net.Expressions/ExpressionComparator.cs [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+using System.IO;
+using Lucene.Net.Index;
+using Lucene.Net.Queries.Function;
+using Lucene.Net.Search;
+using Lucene.Net.Support;
+
+namespace Lucene.Net.Expressions
+{
+	/// <summary>A custom comparator for sorting documents by an expression</summary>
+	internal class ExpressionComparator : FieldComparator<double>
```
Comment (laimis) [inline src/Lucene.Net.Expressions/ExpressionComparator.cs]: Lucene version is using Double instead of double, which can be null. Do we want to keep it that way here too by using double? to match what Lucene has? It is hard to tell for me if it ever deals with nulls, but something to consider.

File: src/Lucene.Net.Expressions/ExpressionComparator.cs [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+		        var context = new Dictionary<string, object>();
+		        Debug.Assert(value != null);
+		        context["scorer"] = value;
+		        scores = source.GetValues(context, readerContext);
+		    }
+		}
+
+		public override int Compare(int slot1, int slot2)
+		{
+            return values[slot1].CompareTo(values[slot2]);
```
Comment (laimis) [inline src/Lucene.Net.Expressions/ExpressionComparator.cs]: indentation

File: src/Lucene.Net.Expressions/ExpressionRescorer.cs [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+			, int docID)
+		{
+			Explanation result = base.Explain(searcher, firstPassExplanation, docID);
+			IList<AtomicReaderContext> leaves = searcher.IndexReader.Leaves;
+			int subReader = ReaderUtil.SubIndex(docID, leaves);
+			AtomicReaderContext readerContext = leaves[subReader];
+			int docIDInSegment = docID - readerContext.DocBase;
+			var context = new Dictionary<string, object>();
+			var fakeScorer = new FakeScorer {score = firstPassExplanation.Value, doc = docIDInSegment};
+		    context["scorer"] = fakeScorer;
```
Comment (laimis) [inline src/Lucene.Net.Expressions/ExpressionRescorer.cs]: indentation

File: msbuild.log (hunk 1 above)
Comment (hakeemsm) [inline msbuild.log]: I can't find this file in my local or within the sln. Not sure where it is coming from and the Config\Platform is set to Debug\Any CPU

**Comment by hakeemsm** [comment]
OK, all issues are fixed except for the first. I don't see the file on my machine.

The issue in Facet may have been caused when I was trying to fix the naming issue on the master branch. Thx

**Comment by hakeemsm** [comment]
Hey, All items are fixed now including the sln file. Thx

File: src/Lucene.Net.TestFramework/Util/LuceneTestCase.cs [hunk 9]
```diff
@@ -974,9 +974,7 @@ public static AlcoholicMergePolicy NewAlcoholicMergePolicy(Random r, TimeZone tz
 
         public static LogMergePolicy NewLogMergePolicy(Random r)
         {
-            // LUCENENET TODO: don't use LogByteSizeMergePolicy until it is fixed
-            //LogMergePolicy logmp = r.NextBoolean() ? (LogMergePolicy)new LogDocMergePolicy() : new LogByteSizeMergePolicy();
-            LogMergePolicy logmp = new LogDocMergePolicy();
+            LogMergePolicy logmp = r.NextBoolean() ? (LogMergePolicy)new LogDocMergePolicy() : new LogByteSizeMergePolicy();
```
Comment (laimis) [inline src/Lucene.Net.TestFramework/Util/LuceneTestCase.cs]: Please put back the code that was there before as LogByteSizeMergePolicy is not behaving properly at the moment. in certain scenarios.

**Comment by synhershko** [comment]
I moved ahead and just copied the files over, way easier

Any further work on Expressions (e.g. fixing tests) needs to happen on a branch and merged via a clean PR

Thanks again @hakeemsm !

**Comment by hakeemsm** [comment]
OK, will do. I got more tests in the green now! Thanks
